File: solutions/229_majority_element_ii.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <map>
#include <stdio.h>
#include <unordered_map>
#include <vector>
/**
Given an integer array of size n, find all elements that appear more than
&lfloor; n/3 &rfloor; times.
Note: The algorithm should run in linear time and in O(1) space.
Example 1:
Input: [3,2,3]
Output: [3]
Example 2:
Input: [1,1,1,3,3,2,2,2]
Output: [1,2]

**/
using namespace std;

static int speed_up = []() {
  std::ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

class Solution {
public:
  vector<int> majorityElement(vector<int> &nums) {
    vector<int> ans;

    int can1 = 0, cter1 = 0, can2 = 1, cter2 = 0;
    for (int n : nums) {
      if (n == can1)
        cter1++;
      else if (n == can2)
        cter2++;
      else if (cter1 == 0) {
        can1 = n;
        cter1 = 1;
      } else if (cter2 == 0) {
        can2 = n;
        cter2++;
      }

      else {
        cter1--;
        cter2--;
      }
    }

    cter1 = cter2 = 0;
    int bound = nums.size() / 3;

    for (int n : nums) {
      if (n == can1)
        cter1++;

      if (n == can2)
        cter2++;

      if (cter1 > bound) {
        ans.push_back(can1);
        cter1 -= nums.size();
      }
      if (cter2 > bound) {
        ans.push_back(can2);
        cter2 -= nums.size();
      }
    }
    return ans;
  }
};

int main() {
  Solution s;
  vector<int> nums{3, 0, 3, 4}; // 1, 2, 2, 1, 2, 3, 1, 3};
  for (int n : s.majorityElement(nums))
    cout << n << endl;
}

File: _site/solutions/229_majority_element_ii.rb
# Given an integer array of size n, find all elements that appear more than &lfloor; n/3 &rfloor; times.
# Note: The algorithm should run in linear time and in O(1) space.
# Example 1:
# Input: [3,2,3]
# Output: [3]
# Example 2:
# Input: [1,1,1,3,3,2,2,2]
# Output: [1,2]
#

# @param {Integer[]} nums
# @return {Integer[]}
def majority_element(nums)
  cache = Hash.new(0).tap { |h| nums.each { |i| h[i] += 1 } }
  sen = nums.size / 3
  cache.keys.keep_if { |k| cache[k] > sen }
end

def majority_element2(nums)
  cana = canb = ctera = cterb = 0
  nums.each do |n|
    if n == cana
      ctera += 1
    elsif n == canb
      cterb += 1
    elsif ctera.zero?
      ctera = 1
      cana = n
    elsif cterb.zero?
      cterb = 1
      canb = n
    else
      ctera -= 1
      cterb -= 1
    end
  end
  [cana, canb].keep_if { |c| nums.count(c) > nums.size / 3 }.uniq
end

nums = [1, 1, 1, 2, 2, 3, 2, 3]
nums = [3, 0, 3, 4]
nums = [0, 0, 0]
p majority_element(nums)
p majority_element2(nums)
